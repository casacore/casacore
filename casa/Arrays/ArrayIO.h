//# ArrayIO.h: text output and binary IO for an array of any dimensionality.
//# Copyright (C) 1993,1994,1995,1997,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ARRAYIO_H
#define CASA_ARRAYIO_H

//# Includes
#include <casa/aips.h>
#include <casa/BasicSL/String.h>

//# Forward declarations
#include <casa/iosfwd.h>

namespace casa { //# NAMESPACE CASA - BEGIN

class AipsIO;
class LogIO;
class IPosition;
template<class T> class Array;
template<class T> class Matrix;
template<class T> class Vector;
template<class T> class Cube;
template<class T> class Block;

// <summary>
//    Input/output operators for Arrays.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Paul Shannon" date="1995/02/21" tests="" demos="">
// This header was reviewed and revised with the goal of making it an
// example for those writing global function header files.
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Array>Array</linkto>
//   <li> ostream
//   <li> <linkto class=AipsIO>AipsIO</linkto>
// </prerequisite>

// <etymology>
//  ArrayIO is simply the conventional shorthand for "array input/output".
// </etymology>

// <synopsis> 
// These global functions provide easy input and output of (possibly)
// large and (possibly) multi-dimensional arrays.  Iteration through
// entire arrays is done behind the scenes, with no effort required
// of the client programmer.  
// These functions are global, rather than member functions of the
// Array class, because of the well-known C++ requirement that the first
// argument to an operator function (as it is declared) is the 
// left operand when the function is called.
// </synopsis>


// <example>
// <srcblock>
// IPosition shape (3,10,10,3);
// Array <Float> array (shape);
// // ...initialize and manipulate the array...
// cout << "result: " << array;
// </srcblock>
//
// <motivation>
// Effortless input/output is clearly a big win.
// </motivation>
//
// <todo asof="1997/01/15">
// </todo>

// <linkfrom anchor="Array IO" classes="Array Vector Matrix Cube">
//    <here>Array IO</here> -- Input/output operators for Arrays.
// </linkfrom>
//
// <group name="Array IO">


// Write out an ascii representation of an array of any dimensionality.
// Arrays of dimensionality 3 or greater are written out vector by vector,
// preceeded by the position of the start of the vector. If the origin of
// the array isn't zero it is printed. The shape of the array is always
// printed.

template<class T> ostream &operator << (ostream &, const Array<T> &);

// Write a formatted copy of the array to the LogIO output object. Merely calls
// the ostream operator<< in turn.
template<class T> LogIO &operator<<(LogIO &os, const Array<T> &a);

// Read an ascii representation of an array. All types with an <src><<</src>
// operator can be handled. The basic format of the input should be:
// <srcblock>
//	[element element element ....] 
// </srcblock>
// Elements are separated by whitespace, or a comma, optionally surrounded
// by white space. <br>
// <note role=warning> Some input routines read fields between blank spaces. This
// is (at the moment) especially true for Quantities and Strings.
// In those cases
// the separator should be blank (or a comma following a blank), and the
// end ']' should have a blank in front.
// A crude fix for String arrays having separators <src>,</src> and <src>]</src>
// without blanks preceding has been made; but slows routines down </note>
// The default input is a vector of unspecified length. The input shape
// can be changed by pre-pending the input with:
// <srcblock>
//	{[shape]}
// </srcblock>
// where shape is an unsigned integer vector. The shape will be used to check
// the input length; and, depending on the possibility, to resize/reshape the
// result. However, reshaping of e.g. a Vector to a Matrix cannot be done, and
// the result will stay in the form asked.<br>
// Input order is row major, however by preceding the input with:
// <srcblock>
//	{T[shape]}	
// </srcblock>
// the order will be reversed.<br>
// Reshaping of the Array provided will depend on the type of Array and its
// state. If a general Array, the shape will be
// as defined by user. If fixed Array (e.g. Matrix, Vector, Cube) the number
// of dimesnsions will be kept. If the user specified more dimensions
// then supported (e.g. 3 for Matrix), the last dimesions will be collapsed.
// If less dimensions are specified, the missing ones will be set to 1.
// will be kept.<br>
// The read() version can be used to force a shape (ip), or an input
// transpose (it) (which can be undone by the user specifying transpose).
//
// <group>

template<class T> istream &operator >> (istream &s, Array<T> &x);
template<class T> Bool read(istream &s, Array<T> &x,
			    const IPosition *ip=0, Bool it=False);
// </group>

// General read support function for matrices.
// In principle these functions will not be
// used by general user, but could be. They can be used by Array type
// classes (like Slice, Lattice) to do the work of comparable input
// functions as the one for Arrays.
// In these functions p is the shape
// of the returned Block x. This shape is either deduced from the user
// specification; made equal to (1, nelements) if no user shape is
// given; is set to ip if specified. The function will return False (and
// p = (0)) in the case of an invalid input element; a number of elements
// input not equal to ip (if specified); the shape given by user as input
// does not conform to ip (if given) or the number of elements input.<br>
// trans will be True if transpose asked by user; or if forced by it.


template<class T> Bool readArrayBlock(istream &s, Bool &trans,
				      IPosition &p,
				      Block<T> &x,
				      const IPosition *ip=0, Bool it=False);


// Read or write a binary representation of an Array to a file. Very
// useful for saving arrays and restoring them later.
// <br>The putArray function is put in for forwards compatibility
// of images (so new images can be read with old release).
//
// <group>

template<class T> AipsIO &operator<< (AipsIO &, const Array<T> &);
template<class T> void putArray (AipsIO &, const Array<T> &, const Char* name);
template<class T> AipsIO &operator>> (AipsIO &, Array<T> &);

// </group>

// </group>




// <summary>
// Global functions to read/write binary arrays from/to a file.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Gareth Hunt" date="95Mar31" tests="" demos="">
// </reviewed>

// <synopsis>
// These global functions provide disk read/write functions for an Array of
// binary numbers.  The write operation is useful, for example, to dump an
// image in binary form to disk so that it can be displayed with an external
// utility such as SAOimage.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<Float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    write_array (picture, fileName);
// </srcblock>
// </example>

// <todo asof="">
//   <li> These functions should eventually be replaced with something
//        more sophisticated.
// </todo>

// <linkfrom anchor="Array binary IO" classes="Array Vector Matrix Cube">
//    <here>Array binary IO</here> -- Simple binary input/output for Arrays.
// </linkfrom>

// <group name=Array binary IO>

// Write the values of an array in binary format into a file with
// the given name.
// The values are stored in local format, thus are not converted
// to a canonical format as
// <linkto class="AipsIO:description">AipsIO</linkto>
// does.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <class T>
void write_array (const Array<T>& the_array, const String& fileName);

template <class T>
inline void write_array (const Array<T>& the_array, const Char* fileName)
    { write_array (the_array, String(fileName)); }
// </group>

// Read the values of an array in binary format from a file with
// the given name.
// The number of values read is the size of the Array, thus the file
// should at least contain that number of values.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <class T>
void read_array (Array<T>& the_array, const String& fileName);

template <class T>
inline void read_array (Array<T>& the_array, const Char* fileName)
    { read_array (the_array, String(fileName)); }
// </group>

// </group>



// <summary>
// Global functions for Matrix/Vector input/output using ASCII format.
// </summary>

// <use visibility=export>

// <prerequisite>
//   <li> <linkto class=Matrix>Matrix</linkto>
//   <li> <linkto class=Vector>Vector</linkto>
// </prerequisite>

// <synopsis>
// These global functions support file I/O between ASCII files and
// Matrices or Vectors.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<Float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    writeAsciiMatrix (picture, fileName);
// </srcblock>
// </example>

// <linkfrom anchor="Array Ascii IO" classes="Vector Matrix">
//    <here>Array Ascii IO</here> -- Simple Ascii input/output for Arrays.
// </linkfrom>

// <group name=Array Ascii IO>

// These routines read and write a Matrix of data.  The first line of
// input will be examined to determine the number of columns in the matrix.
// The maximum number of columns provided for is 100.  Each item may be up 
// to 50 characters long.
//
// Each item must be separated from others by one (or more) blank column.
// The "line" may be up to 1024 characters long.  Each subsequent line must
// contain the SAME number of items as the first line but may be any length
// (up to 1024 characters).
//
// The matrix need NOT be square.  
//
// The matrix should be declared but NOT dimensioned in the calling program.

// <group>
template <class T>
void readAsciiMatrix (Matrix<T>& mat, const Char* fileName);

template <class T>
void writeAsciiMatrix (const Matrix<T>& mat, const Char* fileName);
// </group>


// These two functions read and write a Vector of data.  The input
// may be arranged in any format (i.e. It may be recorded as one value per
// line or it may be recorded with all values on a single line).
// Values must be separated by whitespace.

// <group>
template <class T>
void readAsciiVector (Vector<T>& vec, const Char* fileName);

template <class T>
void writeAsciiVector (const Vector<T>& vec, const Char* fileName);
// </group>

// </group>



} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casa/Arrays/ArrayIO.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
