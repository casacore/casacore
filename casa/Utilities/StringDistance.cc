//# StringDistance.cc: Class to deal with Levensthein distance of strings
//# Copyright (C) 2009
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Utilities/StringDistance.h>

using namespace std;

namespace casa {

StringDistance::StringDistance()
  : itsMaxDistance     (0),
    itsCountSwaps      (False),
    itsIgnoreBlanks    (False),
    itsCaseInsensitive (False)
{}

StringDistance::StringDistance (const String& source, int maxDistance,
                                Bool countSwaps, Bool ignoreBlanks,
                                Bool caseInsensitive)
  : itsSource          (source),
    itsMaxDistance     (maxDistance),
    itsCountSwaps      (countSwaps),
    itsIgnoreBlanks    (ignoreBlanks),
    itsCaseInsensitive (caseInsensitive)
{
  if (ignoreBlanks) {
    itsSource = removeBlanks (itsSource);
  }
  if (caseInsensitive) {
    itsSource.downcase();
  }
  if (itsMaxDistance < 0) {
    itsMaxDistance = 1 + itsSource.size() / 3;
  }
  // Size the matrix such that it suffices for all possible strings.
  itsMatrix.resize (itsSource.size() + 1, itsSource.size() + itsMaxDistance + 1);
  itsMatrix = -1;
}

Bool StringDistance::match (const String& target) const
{
  String t(target);
  if (itsIgnoreBlanks) {
    t = removeBlanks (target);
  }
  int diff = t.size();
  diff -= itsSource.size();
  if (std::abs(diff) > itsMaxDistance) {
    return false;
  }
  if (itsCaseInsensitive) {
    t.downcase();
  }
  if (itsMaxDistance == 0) {
    return t == itsSource;
  }
  return doDistance(itsSource, t, itsCountSwaps, itsMatrix) <= itsMaxDistance;
}

Int StringDistance::distance (const String& target) const
{
  String t(target);
  if (itsIgnoreBlanks) {
    t = removeBlanks (target);
  }
  if (t.size() > itsSource.size() + itsMaxDistance) {
    return t.size() - itsSource.size();
  }
  if (itsCaseInsensitive) {
    t.downcase();
  }
  return doDistance (itsSource, t, itsCountSwaps, itsMatrix);
}

Int StringDistance::distance (const String& source, 
                              const String& target,
                              Bool countSwaps)
{
  Matrix<Int> matrix (source.size() + 1, target.size() + 1);
  return doDistance (source, target, countSwaps, matrix);
}

Int StringDistance::doDistance (const String& source,
                                const String& target,
                                Bool countSwaps,
                                Matrix<Int>& matrix)
{
  int n = source.size();
  int m = target.size();
  if (n == 0) {
    return m;
  }
  if (m == 0) {
    return n;
  }
  // Initialize the first row and column.
  for (int i=0; i<=n; ++i) {
    matrix(i,0) = i;
  }
  for (int j=0; j<=m; ++j) {
    matrix(0,j) = j;
  }
  // Loop over all characters in source and target.
  for (int j=0; j<m; ++j) {
    char t_j = target[j];
    for (int i=0; i<n; ++i) {
      char s_i = source[i];
      // If equal, no cost involved.
      int cost = (s_i == t_j  ?  0 : 1);
      // Step 6
      int above = matrix(i,j+1);
      int left  = matrix(i+1,j);
      int diag  = matrix(i,j);
      int cell  = std::min(above + 1, std::min(left + 1, diag + cost));
      // Step 6A: Cover transposition, in addition to deletion,
      // insertion and substitution. This step is taken from:
      // Berghel, Hal ; Roach, David : "An Extension of Ukkonen's 
      // Enhanced Dynamic Programming ASM Algorithm"
      // (http://www.acm.org/~hlb/publications/asm/asm.html)
      if (countSwaps && i>0 && j>0) {
        int trans = matrix(i-1,j-1) + 1;
        if (source[i-1] != t_j) trans++;
        if (s_i != target[j-1]) trans++;
        if (cell > trans) cell=trans;
      }
      matrix(i+1,j+1) = cell;
    }
  }
  // Now the last element contains the distance.
  return matrix(n,m);
}

String StringDistance::removeBlanks (const String& source)
{
  String dest;
  int n = source.size();
  dest.reserve (n);
  for (int i=0; i<n; ++i) {
    if (source[i] != ' ') {
      int sz = dest.size();
      dest.resize (sz+1);
      dest[sz] = source[i];
    }
  }
  return dest;
}

} //# end namespace


/*
http://www.berghel.net/publications/asm/asm.php
p := k;
   repeat
       inc := p;
       for temp_p := 0 to p-1
           if |(n-m) - inc| <= temp_p
               f((n-m)-inc,temp_p);
         endif
         if |(n-m) + inc| <= temp_p
             f((n-m)+inc,temp_p);
         endif
         inc := inc - 1;
       endfor
       f(n-m,p);
       p := p + 1;
   until FKP[(n-m)+ZERO_K,p-1] = m;
   s := p-1;
   procedure f(k, p)
   begin
       t := FKP[k+ZERO_K,p-1] + 1;
       t2 := t;
       if atat+1 = bk+t+1bk+t then
           t2 := t+1;
       t := max(t, FKP[k-1+ZERO_K,p-1],
           FKP[k+1+ZERO_K,p-1] + 1,t2);
       while at+1 = bt+1+k and t < min(m,n-k) do
           t := t+1;
       FKP[k+ZERO_K,p] := t;
   end
*/
