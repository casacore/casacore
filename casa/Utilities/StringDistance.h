//# StringDistance.h: Class to deal with Levensthein distance of strings
//# Copyright (C) 2009
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_STRINGDISTANCE_H
#define CASA_STRINGDISTANCE_H

//# Includes
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/Arrays/Matrix.h>

namespace casa {

// <summary>
// Class to deal with Levensthein distance of strings.
// </summary>

// <synopsis>
// The Levenshtein Distance is a metric telling how similar strings are.
// It is also known as the Edit Distance.
//
// The distance tells how many operations (i.e., character substitutions,
// insertions, and deletions are needed to transform one string into another.
// <br>There are several extensions to the basic definition:
// <ul>
//  <li> Treat a swap of adjacent characters as one operation.
//  <li> Give a weight to operations (e.g., insertions and deletions have
//       half the weight of the other operations).
//  <li> Take locality into account. For example, successive substitutions
//       weigh more than non-successive.
//  <li> Extend to wildcarded strings.
// </ul> 
// This class optionally uses the swap extension. Furthermore one can
// optionally ignore blanks. By default both options are used.
//
// The code is based on code written by Anders Sewerin Johansen.
// Calculating the distance is an expensive O(N^2) operation, thus
// should be used with care.
//
// The class is constructed with the source string to compare against.
// Thereafter its <code>match</code> or <code>distance</code>
// function can be used for each target string.
// </synopsis>

class StringDistance
{
public:
  // Default constructor sets maxDistance to 0.
  StringDistance();

  // Construct from the source string and maximum distance.
  // If the maximum distance is negative, it defaults to 1+strlength/3.
  // Note that maximum distance 0 means that the strings must match exactly.
  explicit StringDistance (const String& source, Int maxDistance=-1,
                           Bool countSwaps=True, Bool ignoreBlanks=True,
                           Bool caseInsensitive=False);

  // Get data members.
  // <group>
  const string& source() const
    { return itsSource; }
  Int maxDistance() const
    { return itsMaxDistance; }
  const Matrix<Int>& matrix() const
    { return itsMatrix; }
  // </group>
  
  // Test if the given target string is within the maximum distance.
  Bool match (const String& target) const;

  // Calculate the distance from the string to the string given in the constructor.
  // If the length of target exceeds source length + maxDistance,
  // the difference in lengths is returned.
  Int distance (const String& target) const;

  // Calculate the distance between the two strings.
  // This is slower than the <src>distance</src> member function, because
  // it has to allocate the underlying Matrix for each invocation.
  static Int distance (const String& source, const String& target,
                       Bool countSwaps=True);

  // Remove blanks from the given string.
  static String removeBlanks (const String& source);

private:
  // Calculate the distance.
  static Int doDistance (const String& source, const String& target,
                         Bool countSwaps, Matrix<Int>& matrix);


private:
  String              itsSource;
  mutable Matrix<Int> itsMatrix;
  Int                 itsMaxDistance;
  Bool                itsCountSwaps;
  Bool                itsIgnoreBlanks;
  Bool                itsCaseInsensitive;
};

} //# end namespace

#endif
