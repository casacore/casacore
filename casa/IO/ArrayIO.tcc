//# ArrayIO.cc: text output and binary IO for an array of any dimensionality.
//# Copyright (C) 1993,1994,1995,1996,1997,1999,2000,2001,2002,2003
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: casa-feedback@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA

#ifndef CASA_ARRAYIO_2_TCC
#define CASA_ARRAYIO_2_TCC

//# Includes
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Arrays/ArrayError.h>
#include <casacore/casa/Arrays/Matrix.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/ArrayPosIter.h>
#include <casacore/casa/Arrays/ArrayUtil.h>

#include <casacore/casa/IO/AipsIO.h>
#include <casacore/casa/IO/AipsIOCarray.h>
#include <casacore/casa/IO/ArrayIO.h>

#include <casacore/casa/Logging/LogIO.h>

#include <istream>
#include <fstream>
#include <cassert>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

template<class T>
LogIO &operator<<(LogIO &os, const Array<T> &a)
{
    os.output() << a;
    return os;
}

template<class T>
AipsIO &operator<<(AipsIO &ios, const Array<T> &a)
{
    putArray (ios, a, "Array");
    return ios;
}

template<class T>
void putArray (AipsIO &ios, const Array<T> &a, const Char* name)
{
    if (a.size() * sizeof(T) > 2147483647) {
      throw AipsError("AipsIO putArray too large (exceeds 2**31 bytes)");
    }
    ios.putstart(name, Array<T>::arrayVersion());
    // Write out dimensionality
    ios << uInt(a.ndim());
    // Write out length
    for (size_t i=0; i < a.ndim(); i++) {
      ios << uInt(a.shape()(i));
    }
    // Now write out the data
    bool deleteIt;
    const T *storage =  a.getStorage(deleteIt);
    putAipsIO (ios, a.nelements(), storage);
    a.freeStorage(storage, deleteIt);
    ios.putend();
}

// <thrown>
//     <item> ArrayError
// </thrown>
template<class T>
AipsIO &operator>>(AipsIO &ios, Array<T> &a)
{
    // Makes the argument unique, i.e. existing refs will be lost. At the moment
    // this is relatively inefficient as it makes a temporary copy.
    a.unique();
    
    // On 20-Nov-2000 use of the home-brew rtti was removed.
    // It meant that a name 'Array<int>' is now replaced by 'Array'.
    // In order to recognize those old names, we must do something special.
    String type = ios.getNextType();
    int version;
    if (type.length() > 6  &&  type.index("Array<") == 0) {
      version = ios.getstart(type);
    } else {
      version = ios.getstart("Array");
    }
    int ndim;
    ios >> ndim;
    IPosition shape(ndim);
    // Older versions contain an origin (which we discard).
    if (version < 3) {
	int orig;
	for (int i=0; i < ndim; i++) {
	    ios >> orig;
	}
    }
    uInt v;
    for (int i=0; i < ndim; i++) {
      ios >> v;
      shape(i) = v;
    }
    a.resize(shape);                // hopefully a no-op if unchanged

    // Now read in the data.

    bool deleteIt;
    T *storage = a.getStorage(deleteIt);

    uInt nwritten;
    ios >> nwritten;
    if (nwritten != a.nelements())
	throw(ArrayError("AipsIO &operator>>(AipsIO, Array<T>"
			  " - nelements() differs from number in file"));
    getAipsIO (ios, nwritten, storage);
    a.putStorage(storage, deleteIt);
    ios.getend();
    return ios;
}

inline LogIO& operator<< (LogIO& os, const IPosition& ip)
{
    os.output() << ip;
    return os;
}

template<typename T>
Block<T> makeBlock(const Array<T>& array)
{
	Block<T> block(array.nelements());
	if(array.contiguousStorage())
		std::copy(array.cbegin(), array.cend(), block.storage());
	else
		std::copy(array.begin(), array.end(), block.storage());
	return block;
}

template<typename T>
Vector<T> makeVector(const Block<T>& block)
{
	return Vector<T>(block.begin(), block.end());
}

inline Vector<String> stringToVector (const String& string, char delim)
{
	Vector<std::string> vec = strToVector(string, delim);
	return Vector<String>(vec.begin(), vec.end());
}

inline Vector<String> stringToVector (const String& string, const std::regex& delim)
{
	Vector<std::string> vec = strToVector(string, delim);
	return Vector<String>(vec.begin(), vec.end());
}

} //# NAMESPACE CASACORE - END


#endif
