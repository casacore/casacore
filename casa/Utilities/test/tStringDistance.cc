//# tStringDistance.cc: Test of class StringDistance
//# Copyright (C) 2009
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/casa/Utilities/StringDistance.h>
#include <iostream>
#include <stdexcept>
#include <casacore/casa/Arrays/ArrayIO.h>

using namespace std;
using namespace casacore;

int main(int argc, char* argv[2])
{
  try {
    if (argc < 3) {
      cout << "Run as: tStringDistance source target [countSwap] "
           << "[ignoreBlanks] [caseInsensitive]"
           << endl;
      return 1;
    }
    Bool countSwap       = (argc > 3  &&  *argv[3] == '1');
    Bool ignoreBlanks    = (argc > 4  &&  *argv[4] == '1');
    Bool caseInsensitive = (argc > 5  &&  *argv[5] == '1');
    StringDistance dist(argv[1], -1, countSwap, ignoreBlanks, caseInsensitive);
    cout << dist.source() << ' ' << dist.maxDistance() << ": match="
         << dist.match(argv[2]) << " dist=" << dist.distance(argv[2])
         << " distlr=" << StringDistance::distance(argv[1], argv[2], countSwap)
         << " distrl=" << StringDistance::distance(argv[2], argv[1], countSwap)
         << endl;
    cout << dist.matrix();
  } catch (exception& x) {
    cout << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  return 0;
}
