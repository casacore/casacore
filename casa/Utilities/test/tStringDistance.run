#!/bin/sh
#-------------------------------------------------------------------
# Script to test the StringDistance class.
#======================================================

./tStringDistance abcd abcd 0 0 0
./tStringDistance abcd abcd 1 1 1
./tStringDistance AbCd aBcD 0 0 0
./tStringDistance AbCd aBcD 0 0 1
./tStringDistance AbCd abcd 0 0 1
./tStringDistance ACbd abcd 0 0 0
./tStringDistance ACbd abcd 0 0 1
./tStringDistance ACbd abcd 1 0 1
./tStringDistance ACbd 'ab cd' 1 0 1
./tStringDistance ACbd 'ab cd' 1 1 1
./tStringDistance ab ba 1 1 1
