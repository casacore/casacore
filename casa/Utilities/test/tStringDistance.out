abcd 2: match=1 dist=0 distlr=0 distrl=0
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 0, 1, 2, -1, -1
 3, 2, 1, 0, 1, -1, -1
 4, 3, 2, 1, 0, -1, -1]
abcd 2: match=1 dist=0 distlr=0 distrl=0
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 0, 1, 2, -1, -1
 3, 2, 1, 0, 1, -1, -1
 4, 3, 2, 1, 0, -1, -1]
AbCd 2: match=0 dist=4 distlr=4 distrl=4
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 1, 2, 3, 4, -1, -1
 2, 2, 2, 3, 4, -1, -1
 3, 3, 3, 3, 4, -1, -1
 4, 4, 4, 4, 4, -1, -1]
abcd 2: match=1 dist=0 distlr=4 distrl=4
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 0, 1, 2, -1, -1
 3, 2, 1, 0, 1, -1, -1
 4, 3, 2, 1, 0, -1, -1]
abcd 2: match=1 dist=0 distlr=2 distrl=2
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 0, 1, 2, -1, -1
 3, 2, 1, 0, 1, -1, -1
 4, 3, 2, 1, 0, -1, -1]
ACbd 2: match=0 dist=3 distlr=3 distrl=3
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 1, 2, 3, 4, -1, -1
 2, 2, 2, 3, 4, -1, -1
 3, 3, 2, 3, 4, -1, -1
 4, 4, 3, 3, 3, -1, -1]
acbd 2: match=1 dist=2 distlr=3 distrl=3
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 1, 1, 2, -1, -1
 3, 2, 1, 2, 2, -1, -1
 4, 3, 2, 2, 2, -1, -1]
acbd 2: match=1 dist=1 distlr=3 distrl=3
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 1, 1, 2, -1, -1
 3, 2, 1, 1, 2, -1, -1
 4, 3, 2, 2, 1, -1, -1]
acbd 2: match=0 dist=3 distlr=4 distrl=4
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, 5, -1
 1, 0, 1, 2, 3, 4, -1
 2, 1, 1, 2, 2, 3, -1
 3, 2, 1, 2, 3, 3, -1
 4, 3, 2, 2, 3, 3, -1]
acbd 2: match=1 dist=1 distlr=4 distrl=4
Axis Lengths: [5, 7]  (NB: Matrix in Row/Column order)
[0, 1, 2, 3, 4, -1, -1
 1, 0, 1, 2, 3, -1, -1
 2, 1, 1, 1, 2, -1, -1
 3, 2, 1, 1, 2, -1, -1
 4, 3, 2, 2, 1, -1, -1]
ab 1: match=1 dist=1 distlr=1 distrl=1
Axis Lengths: [3, 4]  (NB: Matrix in Row/Column order)
[0, 1, 2, -1
 1, 1, 1, -1
 2, 1, 1, -1]
